# Bike Rental in Seoul

KATE expects your code to define variables with specific names that correspond to certain things we are interested in.

KATE will run your notebook from top to bottom and check the latest value of those variables, so make sure you don't overwrite them.

* Remember to uncomment the line assigning the variable to your answer and don't change the variable or function names.
* Use copies of the original or previous DataFrames to make sure you do not overwrite them by mistake.

You will find instructions below about how to define each variable.

Once you're happy with your code, upload your notebook to KATE to check your feedback.

The dataset used in this assignment was collected to help predict the demand for bike rental in Seoul at any given day and time. It can be found in the University of California Irvine [Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/Seoul+Bike+Sharing+Demand).

The goal of this assignment is to preprocess the input variables into a format that makes them most useful for a linear regression model.

To begin with, let's import the necessary libraries and read in the data:

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

plt.rcParams["figure.dpi"] = 120

data = pd.read_csv("data/seoul_bike_data.csv") 

## Initial data inspection

Now that we have the dataset loaded in, let's inspect it with the methods `.info()` and `.head()`. 

Note that the variable `Rented Bike Count` is the target variable we are aiming to predict in this assignment.

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Date                       8760 non-null   object 
 1   Rented Bike Count          8760 non-null   int64  
 2   Hour                       8760 non-null   int64  
 3   Temperature(°C)            8760 non-null   float64
 4   Humidity(%)                8743 non-null   object 
 5   Wind speed (m/s)           8760 non-null   float64
 6   Visibility (10m)           8760 non-null   int64  
 7   Dew point temperature(°C)  8760 non-null   float64
 8   Solar Radiation (MJ/m2)    8760 non-null   float64
 9   Rainfall(mm)               8760 non-null   float64
 10  Snowfall (cm)              8760 non-null   float64
 11  Seasons                    8760 non-null   object 
 12  Holiday                    8760 non-null   object 
 13  Functioning Day            8760 non-null   objec

In [5]:
data.head(25)

Date  Rented Bike Count  Hour  Temperature(°C) Humidity(%)  \
0   01/12/2017                254     0             -5.2     30%-70%   
1   01/12/2017                204     1             -5.5     30%-70%   
2   01/12/2017                173     2             -6.0     30%-70%   
3   01/12/2017                107     3             -6.2     30%-70%   
4   01/12/2017                 78     4             -6.0     30%-70%   
5   01/12/2017                100     5             -6.4     30%-70%   
6   01/12/2017                181     6             -6.6     30%-70%   
7   01/12/2017                460     7             -7.4     30%-70%   
8   01/12/2017                930     8             -7.6     30%-70%   
9   01/12/2017                490     9             -6.5        <30%   
10  01/12/2017                339    10             -3.5        <30%   
11  01/12/2017                360    11             -0.5        <30%   
12  01/12/2017                449    12              1.7        <30%   
13  01/12/2017                451    13              2.4        <30%   
14  01/12/2017                447    14              3.0        <30%   
15  01/12/2017                463    15              2.1     30%-70%   
16  01/12/2017                484    16              1.2     30%-70%   
17  01/12/2017                555    17              0.8     30%-70%   
18  01/12/2017                862    18              0.6     30%-70%   
19  01/12/2017                600    19              0.0        >70%   
20  01/12/2017                426    20             -0.3        >70%   
21  01/12/2017                405    21             -0.8        >70%   
22  01/12/2017                398    22             -0.9        >70%   
23  01/12/2017                323    23             -1.3        >70%   
24  02/12/2017                328     0             -1.8        >70%   

    Wind speed (m/s)  Visibility (10m)  Dew point temperature(°C)  \
0                2.2              2000                      -17.6   
1                0.8              2000                      -17.6   
2                1.0              2000                      -17.7   
3                0.9              2000                      -17.6   
4                2.3              2000                      -18.6   
5                1.5              2000                      -18.7   
6                1.3              2000                      -19.5   
7                0.9              2000                      -19.3   
8                1.1              2000                      -19.8   
9                0.5              1928                      -22.4   
10               1.2              1996                      -21.2   
11               1.3              1936                      -20.2   
12               1.4              2000                      -17.2   
13               1.6              2000                      -15.6   
14               2.0              2000                      -14.6   
15               3.2              2000                      -11.4   
16               4.2               793                       -7.0   
17               1.6              2000                       -6.5   
18               1.4              2000                       -5.0   
19               1.7              2000                       -3.5   
20               1.5              1913                       -3.5   
21               0.8              1687                       -3.6   
22               1.5              1380                       -3.4   
23               1.0              1265                       -3.6   
24               1.1               994                       -3.6   

    Solar Radiation (MJ/m2)  Rainfall(mm)  Snowfall (cm) Seasons     Holiday  \
0                      0.00           0.0            0.0  Winter  No Holiday   
1                      0.00           0.0            0.0  Winter  No Holiday   
2                      0.00           0.0            0.0  Winter  No Holiday   
3                     

As we can see, there are a lot of different variables which take very different ranges, very different units, and, accordingly, have very different interpretations.

## Categorical vs. Continuous Variables

We must analyse each variable and decide how to preprocess it. Some variables are given in a discrete format, but it makes sense to make them continuous, and vice-versa. 

**Q1. Which variable columns are continuous quantities and which are categorical?**


Create two lists, `categorical` and `continuous` which contain the column indices for the categorical and continuous fields, respectively.

In [4]:
categorical = [0, 2, 4, 11, 12, 13]
continuous = [1, 3, 5, 6, 7, 8, 9, 10]

**Q2. Transform the `Date` variable to a continuous value.**


The `Date` variable could be informative about a trend over time, and can be interpreted as a continuous variable for time. 

Using `.copy()`, create a copy of `data` called `data_date`. 

From `data_date`, create a new variable containing the values of the `Date` column, but converted to a datetime format (using `pd.to_datetime`). Be sure to specify `format="%d/%m/%Y"` otherwise `pandas` may incorrectly parse the dates.

Next, compute **the number of days** between each of these dates and the reference date `01/01/2017` and assign this to a new column in `data_date` called `DayCount`.


Leave the `Date` column unchanged.

*Hint: use the `pd.to_datetime()` function and subtract the values in the `Date` column from the reference date.*

In [1]:
from datetime import datetime

# Add your code below
data_date = data.copy()

date_converted = pd.to_datetime(data_date['Date'], format="%d/%m/%Y")
ref_date = datetime(2017, 1, 1)

day_count = []
for dte in date_converted:
    # dte = date_converted[0]
    day_count.append((dte - ref_date).days)

data_date['DayCount'] = day_count


NameError: name 'data' is not defined

**Q3. Create a new DataFrame with the one-hot representation of the season**


Categorical variables must be formatted to a one-hot representation to be ready for a linear regression model. Create $4$ new columns with the one-hot representation of the Seasons (`Spring`, `Summer`, `Autumn`, and `Winter`, in this order.)

Before starting, create a copy of `data_date` called `data_season` using the `.copy()` method.

*Hint: for each new column we will have rows with values `True` or `False`. To get these values, you can compare the value of the `Seasons` column to that of the new column of interest.*

*In other words the new columns will look like:*

...|Seasons|...|Spring|Summer|Autumn|Winter
---|---|---|---|---|---|---
...|Autumn|...|False|False|True|False
...|Spring|...|True|False|False|False

In [ ]:
data_season = data_date.copy()

data_season["Spring"] = data_season['Seasons'].str.contains('Spring')
data_season["Summer"] = data_season['Seasons'].str.contains('Summer')
data_season["Winter"] = data_season['Seasons'].str.contains('Winter')
data_season["Autumn"] = data_season['Seasons'].str.contains('Autumn')


Once you have implemented the above, uncomment and run the cell below:

In [ ]:
data_season.sample(3)

**Q4. Transform the categorical `Humidity(%)` variable into a continuous quantity.**


In some cases variables are given as discrete categories, but are representing ranges of a continuous values, for example the variable `Humidity(%)`.

Using `.copy()`, create a copy of `data_season` called `data_humidity`. 

Transform the variable `Humidity(%)` to a continuous quantity. Use the mean value of each category as its continuous value (e.g. if a category is "10%-20%", replace it with 15).

If the value of the category is ">70%", use 85 as its mean value. Similarly, if it is "<30%", use 15 as its mean value.

Keep the same column name.

*Hint: You may want to use the `df.replace()` method.*

In [ ]:
data_humidity = data_season.copy()

data_humidity.replace(
    {'30%-70%': sum([30, 70]) / 2, '<30%': sum([0, 30]) / 2, '>70%': sum([70, 100]) / 2}, inplace=True
)
data_humidity["Humidity(%)"].value_counts()

Once you have implemented `data_humidity`, uncomment and run the following cell:

In [ ]:
data_humidity["Humidity(%)"].value_counts()

## Normalisation

Let us inspect the distribution of the continuous variables.

Use the method `.describe()` to produce some basic statistics.

In [ ]:
data.describe()

**Q5. Visualize the distribution of continuous variables.**


Inspect the histograms of the continuous variables in our data using the `.hist()` method. Assign the plot to a variable called `continuous_fig`.

Note that since we have only been altering categorical variables so far, you should use the original `data` DataFrame in this question.

*Hint: use the method `plt.tight_layout()` after calling `.hist()` to create a better arrangement of subplots.*

In [ ]:
# We create a new figure to make other figures in the notebook don't get modified
plt.figure()

# Add your code below
cont_cols = data.describe().columns.tolist()

continuous_fig = data.hist(cont_cols)
plt.tight_layout()


**Q6. Convert continuous variables into categorical variables.**


Note that in the histograms above, these continuous variables have very distinct distribution profiles. For instance, some variables have a very high frequency for a specific value (e.g. zero Solar Radiation). 

It might be useful to have such distinctive values as additional categorical variables.

Create a new DataFrame with four additional columns with the binary variables for values that occur particularly often in the variables `Solar Radiation`, `Snowfall`, `Rainfall` and `Visibility`. 
Create the new columns called `Zero Solar Radiation`, `Zero Snowfall`, `Zero Rainfall` and `Max Visibility`, in this order. 

For instance, the binary variable `Zero Solar Radiation` will indicate if `Solar Radiation` is _close to_ zero (in this question, if it is smaller than 0.1). The same is true for columns `Zero Snowfall` and `Zero Rainfall`.

Since the most common `Visibility` value is not zero, the binary variable `Max Visibility` will indicate if `Visibility` is _close to_ its maximum value (i.e. if its value is greater than its max value - 0.1).

First, however, create a copy of `data_humidity` using the `.copy()` method. Call your new DataFrame `data_binary`.

In [ ]:
data_binary = data_humidity.copy()

data_binary["Zero Solar Radiation"] = [True if x < 0.1 else False for x in data_binary["Solar Radiation (MJ/m2)"]]
data_binary["Zero Snowfall"] = [True if x < 0.1 else False for x in data_binary["Snowfall (cm)"]]
data_binary["Zero Rainfall"] = [True if x < 0.1 else False for x in data_binary["Rainfall(mm)"]]
data_binary["Max Visibility"] = [True if x < 0.1 else False for x in data_binary["Solar Radiation (MJ/m2)"]]

data_binary["Zero Solar Radiation"].value_counts()

Once you have created these new columns, uncomment and run the following cell:

In [ ]:
data_binary["Zero Solar Radiation"].value_counts()

**Q7. Normalise variables.**


It is important to ensure that input data are all scaled to the same range. Without this, the model may produce inaccurate predictions. Normalising each input variable can help train the model, allow easier interpretation of the learned parameters, and offer better regularisation.

Normalise each of the continuous variables in `data` to a *z-scored* DataFrame (such that each column has zero mean and unit variance). Use the transformation:

$$z = \frac{x - \mu}{\sigma}$$

Where `x` is the original column values, `μ` is the mean of the column, and `σ` is the standard deviation of the column.

Save the results in a different DataFrame, `data_z`, and visualise the new distributions using the `.hist()` and `plt.tight_layout()` methods.

Assign the plot to a variable called `normalise_fig`.

*Hint: use the `mean()` and `std()` methods for the chosen columns.*

In [ ]:
# We create a new figure to make other figures in the notebook don't get modified
plt.figure()

# Add your code below
data_z = (data[cont_cols] - data[cont_cols].mean()) / data[cont_cols].std()

normalise_fig = data_z.hist()
plt.tight_layout()


## Advanced discretisation

**Q8. Discretising a circular variable.**


Some continuous variables are not actually linear variables, and are not a natural input to a linear model. For example, `Hour` is a circular variable - the values 23 and 1 are actually close together in the day. 

Let us transform `Hour` into discrete categories.  Divide the variable into 5 categories: `Morning` (6-10), `Afternoon` (11-16), `Evening` (17-19), `Night` (20-23), `Early Morning` (0-5).

For example:

...|Hour|...|Morning|Afternoon|Evening|Night|Early Morning
---|---|---|---|---|---|---|---
...|1|...|False|False|False|False|True
...|13|...|False|True|False|False|False


First, however, create a copy of `data_binary` using the `.copy()` method. Call your new DataFrame `data_time_categories`.

In [ ]:
data_time_categories = data_binary.copy()

data_time_categories["Morning"] = (data_time_categories['Hour'] >= 6) & (data_time_categories['Hour'] <= 10)
data_time_categories["Afternoon"] = (data_time_categories['Hour'] >= 11) & (data_time_categories['Hour'] <= 16)
data_time_categories["Evening"] = (data_time_categories['Hour'] >= 17) & (data_time_categories['Hour'] <= 19)
data_time_categories["Night"] = (data_time_categories['Hour'] >= 20) & (data_time_categories['Hour'] <= 23)
data_time_categories["Early Morning"] = (data_time_categories['Hour'] >= 0) & (data_time_categories['Hour'] <= 5)

data_time_categories["Morning"].value_counts()

Once you have implemented the above question, uncomment and run the following cell:

In [ ]:
data_time_categories["Morning"].value_counts()

**Q9. Inspecting target variable dependency.**


The categories chosen above for `Hour` were rather arbitrary. One more advanced data preprocessing step is to inspect how a given input variable influences the target variable. 

Let's use the `seaborn` library to plot this dependency in detail - how the target variable `Rented Bike Count` depends on `Hour` - with the `sns.violinplot()` method. It shows the distribution of bike rentals for each hour of the day. Assign the output of the plot to a variable called `bike_hour_dependency`. For input data, use the `data` DataFrame.

In [ ]:
# We create a new figure to make other figures in the notebook don't get modified
plt.figure()

# Add your code below
bike_hour_dependency = sns.violinplot(data, x='Hour', y='Rented Bike Count')

We can see that there is complex relation between the variables. Note that as linear models are only sensitive to the mean correlation between variables (not the full distribution), we can focus on the mean values.

**Q10. Calculate the mean `Rented Bike Count` for each hour of the day.**


Use the methods `.groupby()` and `.mean()` on the `data` DataFrame, save the result to a variable called `mean_count`, and plot the result using `.plot()`.

In [ ]:
# Add your code below
mean_count = data[['Hour', 'Rented Bike Count']].groupby('Hour').mean()
mean_count.plot()

**Q11. Discretise nonlinear dependency.**


We see in the plot above, a highly nonlinear effect of the variable `Hour` on the mean of the target variable `Rented Bike Count`. A linear model is not sensitive to this and it is a strong indication that discretisation of the variable will help. As above, let's create categorical variables for different ranges of `Hour`, but now taking into account the dependency of the target variable.

We expect that a category will be more predictive if the target does not vary too much for samples of that category. For instance, the target value is similar for the hours 10, 11, 12 and 13, indicating that the range `10 <= Hour < 14` might have good predictive power for the target.

Divide `Hour` into five new columns as follows:

New column name | Data
---|---
Hour Cat 1 | `3 <= Hour < 7`
Hour Cat 2 | `7 <= Hour < 10`
Hour Cat 3 | `10 <= Hour < 14`
Hour Cat 4 | `14 <= Hour < 22`
Hour Cat 5 | `22 <= Hour` or `Hour < 3`
 
Which will look like:

...|Hour|...|Hour Cat 1|Hour Cat 2|Hour Cat 3|Hour Cat 4|Hour Cat 5
---|---|---|---|---|---|---|---
...|2|...|False|False|False|False|True
...|9|...|False|True|False|False|False

First, create a copy of `data_time_categories` using the `.copy()` method. Call your new DataFrame `final_data`.

In [ ]:
# Add your code below
final_data = data_time_categories.copy()

final_data["Hour Cat 1"] = (3 <= final_data['Hour']) & (final_data['Hour'] < 7)
final_data["Hour Cat 2"] = (7 <= final_data['Hour']) & (final_data['Hour'] < 10)
final_data["Hour Cat 3"] = (10 <= final_data['Hour']) & (final_data['Hour'] < 14)
final_data["Hour Cat 4"] = (14 <= final_data['Hour']) & (final_data['Hour'] < 22)
final_data["Hour Cat 5"] = final_data['Hour'] < 3

Notice that in this case an hour-by-hour division may even be reasonable. But on the other hand, too many input variables can lead to overfitting problems.

## Evaluating predictions

**Q12. Implement linear regression model.**


Create a function with a single argument `cols`, which takes as input a `list` of column names to be used to train the model.

You will need to assign these columns of the `final_data` DataFrame to a variable (`X`), assign the `Rented Bike Count` column to another variable (`Y`), and then use the `.fit()` method of the `LinearRegression` class to train your model.

Note that you will have to reshape your `Y` variable using the `.reshape()` method. It is currently an array (`[254, 204, ...]`) but the `LinearRegression` model expects it as an _array of arrays_ (`[[254], [204], ...]`).

Finally, use the `mean_squared_error` to compute the error. Once computed, return the error as a `float`.

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
# Add your code below
# def prediction_error(cols):
#     ...
#     return error


Uncomment and run the code below to check your implementation.

In [ ]:
# prediction_error(["Temperature(°C)"])

**Q13. Compare performance of different `Hour` representations.**


Compare the predictions of our linear model on the original `Hour` variable versus our hand-crafted categorical representations (`Hour Cat 1`, `Hour Cat 2` etc).

Assign the output of the `prediction_error()` function to two variables:
  1. `model1`: for just the `Hour` variable.
  2. `model2`: for the categorical representations that we created.

Which model has the better performance? Note that with `mean_squared_error`, a lower error indicates a better fit to the data.

In [ ]:
# Add your code below
# model1 = ...
# model2 = ...


**Q14. Compare the prediction error based on the original variables to the prediction error for variables after preprocessing.**

1. Save to the variable `full_model_original`, the result of calling `prediction_error()` on the original columns which are suitable for a linear regression model. In this case these are `['Hour', 'Temperature(°C)', 'Wind speed (m/s)', 'Visibility (10m)', 'Dew point temperature(°C)', 'Solar Radiation (MJ/m2)', 'Rainfall(mm)', 'Snowfall (cm)']`.
<br><br>
2. Save to the variable `full_model_updated`, the result of calling `prediction_error()` using our processed variables. For `full_model_updated`, include the variables from `full_model_original` that we didn't preprocess (like `Temperature(°C)`, `Wind speed (m/s)` etc), but where we have processed a variable (like `Hour`), don't include the unprocessed variable.

In [ ]:
# Add your code below
# full_model_original = ...
# full_model_updated = ...
# print(full_model_original, full_model_updated)
